### PR TITLE
magnifier: disable direct scanout when active

### PR DESCRIPTION
**Pull request by Consolatis**
Fixes:
- #1980

Draft while waiting for confirmation that disabling direct scanout fixes the issue.

### CODE REVIEW

**Comment by johanmalm** [comment]
I haven't looked through `types/scene/wlr_scene.c` properly to see how `scene->direct_scanout` is used, but other than that it seems fine.

**Comment by Consolatis** [comment]
> I haven't looked through `types/scene/wlr_scene.c` properly to see how `scene->direct_scanout` is used, but other than that it seems fine.

At least in 0.17 and 0.18 it is set on `wlr_scene_create()` and then never modified. It is only checked against when deciding if to allow direct scanout or not. As mentioned by @Nefsen402 in https://github.com/labwc/labwc/issues/1980#issuecomment-2231547950 it is marked as private though.

**Comment by Nefsen402** [comment]
The magnifier in labwc sounds like a massive hack to me. It should be implemented by compositing into a staging buffer, then taking that staging buffer and rendering the magnifier. I believe this is possible to do with wlroots without any hacks with 0.18.

Since 0.18, a custom swapchain can be passed to the scene to render into a set of buffers we control and that we will use for creating the magnifier. Here is some pseudo code to achieve this (I'm not going to bother looking up all the exact function names and signatures - I'm just going off of memory):
```c
struct wlr_scene_output *main_scene = ...;

// this output_scene will be replaced with a new scene when the magnifier is active
struct wlr_scene_output *output_scene = main_scene;

if (is_magnifier_used) {
   struct swapchain *schain =  wlr_swapchain_create(output->allocator, output->width, output->height, output->swapchain->format);

   struct wlr_output_state *state;
   wlr_output_state_init(&state);

   wlr_scene_build_output_state(main_scene, &state, &(struct output_options) {
    .swapchain = schain,
  });

  struct wlr_buffer *composited = state->buffer;

  // the composited buffer should then be placed into a secondary scene and the magnifier on top of that.
  output_scene = wlr_scene_create();
  struct wlr_scene_buffer *scene_buffer = wlr_scene_buffer_create(&output_scene->tree);
  wlr_scene_buffer_set_buffer_with_damage(scene_buffer, state->buffer, state->damage);
  labwc_render_maginifier();

  // make sure to free everything
  wlr_output_state_finish(&state);
}

wlr_scene_output_commit(output->scene_output, output_scene);
```

modulo some code to make sure we're not leaking memory and reusing objects like the swapchain.

Basically, how this works is that we have two scenes: Our primary scene that we would normally use like business as usual for the compositor. Then when we activate the magnifier, we create another scene that will handle compositing of the magnifier element after the main compositing has been finished. We use our own custom swapchain to create a place for where staging buffers can live. We can then access the buffer that scene generated for us, and render that into the secondary scene. We can then just use the secondary scene to render the magnifier without worrying about trampling buffer contents we didn't mean to (like with this bug and direct scanout). This should be higher performance as it will never disable direct scanout.

**Comment by Consolatis** [comment]
We need to fix this for wlroots 0.17.x thus we can't rely on anything that requires wlroots 0.18.

Thanks for the pointers, I am completely fine with revisiting the design of the magnifier for labwc 0.8 / wlroots 0.18.
What we currently do is copying the area around the cursor to a temporary wlr_buffer via render_pass, creating a texture out of it and pasting it back with an enlarged dst_box to the output buffer. Then adding the region to the scene output damage ring.

While we do reuse the existing output buffer I would not strictly call this a hack, we are using the wlroots / wlr_scene rendering primitives but missed the direct scanout case that (rightfully I guess) assumes that there can't be any damage on the frame while its active and thus directly maps the client damage to the output damage.

**Comment by johanmalm** [comment]
> Since 0.18, a custom swapchain can be passed to the scene to render into a set of buffers we control and that we will use for creating the magnifier. Here is some pseudo code to achieve this (I'm not going to bother looking up all the exact function names and signatures - I'm just going off of memory):

@Nefsen402 Really appreciate the steer here. That does indeed look a neat way of doing it.

**Comment by johanmalm** [comment]
Patch LGTM. Suggest we merge this to fix the scanout problem.

Thanks for realising what the issue was. I was thinking scanout but when they said "maximized" I stupidly filtered it out being stuck in SSD thinking.

**Comment by Consolatis** [comment]
> It should be implemented by compositing into a staging buffer, then taking that staging buffer and rendering the magnifier. I believe this is possible to do with wlroots without any hacks with 0.18.

What would be the difference (other than one more frame in memory)? The secondary scene would then also need to modify the build_state() buffer (which sounds the same as what we are doing currently) or place another scene node on top for the magnifier (which would disable direct scanout). Do I miss something?

**Comment by Consolatis** [comment]
We do use `wlr_renderer_begin_buffer_pass()` + `wlr_render_pass_add_texture()` with a temporary (but reused within the magnifier) buffer allocated via `wlr_allocator_create_buffer()` with the same format as `state->buffer`. Shouldn't that make it work without CPU copy?

**Comment by Nefsen402** [comment]
Ah, in that case that would be on the GPU. Not a bad solution, also uses less memory. It's just not supported because it violated the ownership model wlroots has for buffers and scene.

**Comment by Nefsen402** [comment]
I wonder if there is a way we can offer a supported way of doing this
